test/controllers/tasks_controller_test.rb

```diff
@@ -1,124 +1,176 @@
   end
-  
+
   # Unskip and complete these tests for Wave 3
-  describe "edit" do
-    it "can get the edit page for an existing task" do
-      skip
-      # Your code here
+  describe 'edit' do
+    it 'can get the edit page for an existing task' do
+      get edit_task_path(task.id)
+
+      must_respond_with :success
+
     end
-    
-    it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+
+    it 'will respond with redirect when attempting to edit a nonexistant task' do
+      get edit_task_path(-1)
+
+      must_respond_with :redirect
     end
   end
-  
+
   # Uncomment and complete these tests for Wave 3
-  describe "update" do
-    # Note:  If there was a way to fail to save the changes to a task, that would be a great
-    #        thing to test.
-    it "can update an existing task" do
-      # Your code here
+  describe 'update' do
+    before do
+      Task.create(name: 'test', description: "It's a test!", completed_at: "test o'clock")
     end
-    
-    it "will redirect to the root page if given an invalid id" do
-      # Your code here
+    let (:new_task_hash) {
+      {
+          task: {
+              name: 'A whole new test!',
+              description: "it's a whole new test!",
+              completed_at: "second test o'clock",
+          },
+      }
+    }
+    it 'can update an existing task' do
+      id = Task.first.id
+      expect {
+        patch task_path(id), params: new_task_hash
+      }.wont_change 'Task.count'
+
+      must_respond_with :redirect
+
+      task = Task.find_by(id: id)
+      expect(task.name).must_equal new_task_hash[:task][:name]
+      expect(task.description).must_equal new_task_hash[:task][:description]
+      expect(task.completed_at).must_equal new_task_hash[:task][:completed_at]
+    end
+
+    it 'will redirect to the root page if given an invalid id' do
+      id = -1
+
+      expect {
+        patch task_path(id), params: new_task_hash
+      }.wont_change 'Task.count'
+
+      must_redirect_to tasks_path
     end
   end
-  
+
   # Complete these tests for Wave 4
-  describe "destroy" do
+  describe 'destroy' do
     # Your tests go here
-    
+    it 'can destroy a task' do
+      task = Task.new(name: 'test', description: 'testy test', completed_at: 'done done done')
+
+      task.save
+      id = task.id
+
+      expect {
+        delete task_path(id)
+      }.must_change 'Task.count', -1
+      must_respond_with :redirect
+      must_redirect_to tasks_path
+    end
   end
-  
+
   # Complete for Wave 4
-  describe "toggle_complete" do
-    # Your tests go here
+  describe 'markup' do
+    it 'will change complete at' do
+      task = Task.new(name: 'test', description: 'testy test', completed_at: '')
+      id = task.id
+
+      task.save
+
+      expect {
+        patch markup_task_path(id)
+
+        expect(task.completed_at).wont_equal ''}
+      # must_respond_with :redirect
+      # must_redirect_to tasks_path
```

Comment:
It works out, but I think technically the syntax you want is this:

```suggestion
      task = Task.new(name: 'test', description: 'testy test', completed_at: '')
      task.save
      id = task.id

      patch markup_task_path(id)

      task = Task.find_by(id: id)
      
      expect(task.completed_at).wont_equal ''
```

1. You should move `id = task.id` below `task.save`. Recall, tasks don't have an ID until they're stored in the database-- so you should get `id` after saving to the database.
2. You'll probably want the typical syntax for making a patch request: `patch markup_task_path(id)`, without the leading `expect {` on line 168.
3. You'll need to find `task` again after the patch request and before the `expect` part, so you can find the task with `task = Task.find_by(id: id)`